src/validations/docs/adr/0011-dns-validation.md

```diff
@@ -0,0 +1,99 @@
+# 11. DNS validation within Schematron
+
+Date: 2022-04-08
+
+## Status
+
+Proposed
+
+## Context
+
+Some validations require DNS queries to be performed from within Schematron.
+
+There are public DNS services which provide DNS over HTTPS (DoH) with responses in JSON format.
+([RFC8484](https://datatracker.ietf.org/doc/html/rfc8484)'s wire format is non-trivial to implement in pure Schematron.)
+
+JSON response format is compatible with the XPath functions `unparsed-text()` and `unparsed-text-available()`
+
+- There is no normative specification for DoH with JSON.
+- Google and Cloudflare have DoH services with equivalent APIs.
+- Both have high availability via anycast.
+- Both are established companies.
+- Both DNS services have been available for several years.
+- Both DNS services are widely used
+- Both DNS services support all DNS RR types.
+
+### Proof of Concept
+
+The following Schematron code sample was used to test DoH with JSON response.
+
+```
+<?xml version="1.0" encoding="UTF-8"?>
+<sch:schema
+    queryBinding="xslt2"
+    xmlns:array="http://www.w3.org/2005/xpath-functions/array"
+    xmlns:map="http://www.w3.org/2005/xpath-functions/map"
+    xmlns:sch="http://purl.oclc.org/dsdl/schematron"
+    xmlns:sqf="http://www.schematron-quickfix.com/validator/process">
+
+    <sch:ns
+        prefix="array"
+        uri="http://www.w3.org/2005/xpath-functions/array" />
+    <sch:ns
+        prefix="map"
+        uri="http://www.w3.org/2005/xpath-functions/map" />
+
+    <sch:pattern
+        id="dnssec">
+
+        <sch:let
+            name="dnssec_href"
+            value="'https://dns.google/resolve?name=fedramp.gov&amp;type=soa'" />
+
+        <sch:let
+            name="dnssec_data"
+            value="unparsed-text($dnssec_href)" />
+
+        <sch:rule
+            context="/">
+
+            <sch:report
+                role="information"
+                test="true()"><sch:value-of
+                    select="$dnssec_data" /></sch:report>
+
+            <sch:let
+                name="response"
+                value="parse-json($dnssec_data)" />
```

Comment:
How many DNS lookups generally would be tested per-SSP? This form of request might be slow in Javascript, given that SaxonJS doesn't natively do async IO. Perhaps using the async version `a:parse-json` (https://www.balisage.net/Proceedings/vol25/print/Kay01/BalisageVol25-Kay01.html) is viable? (not sure if that may be done in a way compatible with the Java runtime)

Reply:
**For validating DNSSEC presence**, the number of queries is proportional to the number of distinct DNS domains and sub-domains. That could be as low as one or a few, but determining how many is not simple. Host names and network addresses (only globally routable network addresses need be checked) in OSCAL SSPs may appear in `<component>` and `<inventory-item>` `<prop>`s and a few other places where DNSSEC is pertinent. Minimizing the number of related DNS queries by identification of the distinct DNS domains and sub-domains (SOAs) is technically feasible but not simple. A simple approach would simply query each (distinct) host name or address to see if DNSSEC is in use. Multiple query targets (_questions_ in the API request) can be bundled into a single DoH API request but the maximum number of questions per API request is likely limited.

A quick search for minimization techniques (name and address mapping to SOAs) was not successful.

For a DNSSEC check in the absence of a handy "How many queries will I need" algorithm we should probably just calculate the number of distinct host names (or even better obvious second- or third-level domains) found in the SSP and limit the query/queries to a fixed number.

Anything else would have to build a tree right to left starting at the root of all host names (i.e., "dot") and use it to check DNSSEC+SOA at each node until it got tired of doing so. An analogous process exists for network addresses. A rough example can be seen [here](https://dnsviz.net/d/fedramp.gov/dnssec/).

Doing that with assuredly terminating asynchronous execution "inside" Schematron (ultimately XSLT) adds difficulty.